AWSAuthSDK/Sources/AWSMobileClient/AWSMobileClientOperations/AWSMobileClient+Tokens.swift

```diff
@@ -77,7 +77,12 @@ extension AWSMobileClient: FetchUserPoolTokensDelegate {
     }
 
     func getCurrentUsername(operation: FetchUserPoolTokensOperation) -> String? {
-        username
+        // このdelegate関数はトークン取得時のみ呼ばれ、saml認証ではCognitoIdentityUserBehavior取得時usernameが使われない模様。
+        // ただusernameがないとTokenが取得される前に弾かれるので、仮のusername（空にならないような値）を返すようにしている   2022.11.22 kim
+        if case .hostedUI = self.federationProvider, username == nil {
+            return "TalknoteOAuth"
```

Comment:
ここダミーで返しちゃって大丈夫なものです？
cognitoに対してwrite的なアクセスがなければ大丈夫なのかな？っていう気もするのですが、usernameはuser_idみたいなもので、cognitoが発番するのでちょっと怖いなっていう気がしました。

Reply:
怖いですね！
ただ調べた結果、write的なアクセスはなさそうです。
ここのusernameはトークン取得時のみ呼ばれていて
https://github.com/talknote-inc/aws-sdk-ios/blob/5aef153412c9e359c122493af7d460a3b76ab7da/AWSAuthSDK/Sources/AWSMobileClient/AWSMobileClientOperations/AWSAsyncOperations/GetTokensUserPool/FetchUserPoolTokensOperation.swift#L121-L132
userPool.getIdentityUserでusernameは使われてないです。
https://github.com/talknote-inc/aws-sdk-ios/blob/5aef153412c9e359c122493af7d460a3b76ab7da/AWSAuthSDK/Sources/AWSMobileClient/Helpers/UserPoolHelper/CognitoUserPoolBehavior.swift#L29-L33

Reply:
色々追えてなくて申し訳ないのですが、教えてください！
なんでsamlの時にはusernameが設定されないんでしょう？
cognitoからSAML側IDPに遷移して、cognitoに戻ってきた時にusernameつかめそうな気がしないでもなく。。。
そもそもaws-sdkってsaml認証に対応してないんでしたっけ？
https://github.com/talknote-inc/aws-sdk-ios/pull/9/files#diff-718247e162f9fd35c1b831a0c9fd691389f540b475b3961fbb427601eb39e2cbR81
このあたりで取れてたりとかってないですかね？
すいません。ちょっとシーケンスも把握しておらず。。。

Reply:
> なんでsamlの時にはusernameが設定されないんでしょう？
cognitoからSAML側IDPに遷移して、cognitoに戻ってきた時にusernameつかめそうな気がしないでもなく。。。
そもそもaws-sdkってsaml認証に対応してないんでしたっけ？

自分も把握しきれてないですが、以下のページを見る限りサポートはされているっぽいですね。
ただAWSIdentityProviderManagerなど直接ただいてないので、使い方はちょっと違っているかもです。
もしかしたらsaml認証でもShareExtensionを使わせる意図があっての実装かもしれないです。
https://docs.aws.amazon.com/ja_jp/cognito/latest/developerguide/saml-identity-provider.html

> https://github.com/talknote-inc/aws-sdk-ios/pull/9/files#diff-718247e162f9fd35c1b831a0c9fd691389f540b475b3961fbb427601eb39e2cbR81
このあたりで取れてたりとかってないですかね？

SAML認証用の情報(GET /api/v1/saml_config)取得で受け取ったclientIdの保存用で使われているみたいで、usernameは持ってないですね。
https://github.com/talknote-inc/talknote_ios_oberon/blob/cece40b019dc80d695be4a99e78a576146ba9ceb/Talknote/DataLayer/Auth/DataStore/RemoteDataStore/CognitoDataStore.swift#L285

ちなみに、今回usernameのは`userpoolOpsHelper.currentActiveUser?.username`で取得されるんですが、試しで以下のコードを入れて確認したところ、usernameがちゃんと存在していました。
```swift
let cognitoAuth = AWSCognitoAuth.init(forKey: AWSMobileClientConstants.CognitoAuthRegistrationKey)
cognitoAuth.getSession { session, _ in
	print("username: \(session?.username)")  // samltalknotetest-page-gsuite_admin@samltalknotetest.page
}
```
ということは`userpoolOpsHelper.currentActiveUser?.username`とAWSCognitoAuthのSessionが持っているusernameの同期がとれていないとかですかね。この辺りもう少し調べます。

Reply:
まずわかったことを書きます。
Token取得の際に、AWSMobileClientは都度FetchUserPoolTokensOperationを生成します。
FetchUserPoolTokensOperationはuserpoolを持っていまして、普通のID/Passログイン時はAWSCognitoIdentityUserPoolが使われ、SAML認証等の際はAWSCognitoAuthが使われます。
それぞれ名前の異なるキーチェーンクラス持ってTokenを管理しています。（keychainにデータ保存時にAWSCognitoIdentityUserPoolとAWSCognitoAuth両方の整合性を保つというコメントも見受けられますが、そもそもclientIdが異なり整合性が保てないという印象です。keychainからのusername取得のキーでclientIdが使われます）

で、今回のusernameはself.userpoolOpsHelper.userpoolClient.currentUserから取得していますが、
ここでのuserpoolClientはAWSCognitoIdentityUserPoolクラスで、saml認証時にusernameが取得されないのは想定の動きかもとも思われます。

※usernameに書いてあるコメント
Returns the username attribute of the access token for the current logged in user,  nil otherwise.
Note that the value stored may vary depending on how sign-in was performed.
（DeepL訳：現在ログインしているユーザのアクセストークンの username 属性を返す。それ以外の場合は nil を返す。
ただし、サインインの方法によっては、格納される値が異なる場合があります。）

で、とにかくusernameが取れないとToken取得時に弾かれるので、saml認証の際はAWSCognitoAuthからusernameを取得するように修正しました。 https://github.com/talknote-inc/aws-sdk-ios/pull/9/commits/aa76b7c56da18e7d720d05c4ed52e5bd08a3d85b
尚、saml認証の際、ShareExtensionで投稿できない事象にも対応しています。 https://github.com/talknote-inc/aws-sdk-ios/pull/9/commits/7dcd4996f88463a3591b2c8034ec6ee8ba0405e9